SMLHelper/Utility/PatchUtils.cs

```diff
@@ -19,5 +23,24 @@ internal static void PatchList<ValueType>(List<ValueType> original, IList<ValueT
                 original.Add(entry);
             }
         }
+
+        // use methods from 'typeWithPatchMethods' class as harmony patches
+        // valid method need to have HarmonyPatch and Harmony[Prefix/Postfix/Transpiler] attributes
+        // if typeWithPatchMethods is null, we use type from which this method is called
+        internal static void PatchClass(HarmonyInstance harmony, Type typeWithPatchMethods = null)
```

Comment:
HarmonyPatch attribute have a lot of code, but Prefix/Postfix/Transpiler attributes are just empty. I can add three new attributes for those so for patch method it will be
```cs
[SomeCustomNameHarmonyPrefix]
[HarmonyPatch(type, name)]
static void method()
```
So in case of any confusion it'll be possible to just go to the definition of that prefix attribute and see this PatchClass method.

Reply:
Okay, I feel like I should weigh in on this. 

Attributes from an assembly should never be used to perform something in another assembly. Like Prime said, this will just cause confusion. You should create custom attributes like:
```cs
[Prefix(type, name)]
[Postfix(type, name)]
[Transpiler(type, name)]
``` 

But scratch that, since there is absolutely no reason to introduce this sort of thing into SMLHelper.

If we wanted to use attribute patching, we could've just used harmony's `PatchAll`.

Reply:
PatchAll don't work with methods, only with classes. So that HarmonyPatch attribute will not be processed by PatchAll even if you use it.

Reply:
After all is said an done. I agree with Alex and my initial assessment. 

All these patch methods can be converted to the basic Harmony Patch method.  
Yes, it may look a little uglier with all of the null parameters, but it would maintain consistency throughout the project without introducing multiple different ways in which each patcher class does its job.

If you need help converting these methods to use the Patch method format, I can help with that.

Reply:
While I understand why patch order for classes can be important (I don't think that maintaining order for all patch classes is necessary though, because that way you can't really say which classes depends on that order), I don't see any reason to keep order for patching methods. Using attributes for that removes the need of something like [that](https://github.com/SubnauticaModding/SMLHelper/blob/22427d0fcb374839582bdacb44b50732ae97a567/SMLHelper/Patchers/CraftTreePatcher.cs#L25-L47)

Reply:
That way you keep information about patches in two different places instead of one

Reply:
> That way you keep information about patches in two different places instead of one

I don't see how that's a benefit.

> Using attributes for that removes the need of something like [that](https://github.com/SubnauticaModding/SMLHelper/blob/22427d0fcb374839582bdacb44b50732ae97a567/SMLHelper/Patchers/CraftTreePatcher.cs#L25-L47)

I don't see this as a problem. In fact, it has no more `typeof` and `nameof` usages than the attribute patching method would.

---

In the end, it's about maintain consistency throughout the problem and mitigating the amount of "code entropy" we already have.  
All the patcher classes should be using the same patterns, with any one of them serving as a template for how to make a new one.  
Introducing "yet another" way to perform a Harmony patch, especially one that is not documented by Harmony themselves, is going to make things messy for us down the road and make things even more complicated for potential new code contributors.

Also, consider that many patchers in SMLHelper perform patching actions that don't necessarily involve Harmony at all.  
For these cases, having these patching "Harmony-like" attributes would just add more confusion.

As I have said before, I am willing to help with converting these new patches to the Patch method format.

Reply:
>> That way you keep information about patches in two different places instead of one

>I don't see how that's a benefit.

Well, in my book, breaking information about one single entity to different places is not quite best practice.

> In fact, it has no more typeof and nameof usages than the attribute patching method would.

You need to duplicate patch method name, that 'postfix' parameter. With attributes, you need to specify only target method.

>All the patcher classes should be using the same patterns, with any one of them serving as a template for how to make a new one.

And that's the reason I added it to common PatchUtils class and not just in OptionsPatcher. You can switch to this over time and it can be done class by class, no need to change it all at once. 

>Also, consider that many patchers in SMLHelper perform patching actions that don't necessarily involve Harmony at all.
>For these cases, having these patching "Harmony-like" attributes would just add more confusion.

Didn't understand that :thinking:. If you don't need to use Harmony then you don't need to use Harmony attributes and don't need to use PatchClass().

>As I have said before, I am willing to help with converting these new patches to the Patch method format.

You saying this for the third time, you think I did it because I don't understand how to use Harmony.Patch? :)